### ISSUE TITLE
bayes_R2 for brmsfit_multiple objects

### ISSUE THREAD

**Issue by jmgirard** [paul-buerkner/brms#997]
When I try to use `bayes_R2()` on a `brmsfit_multiple` object, I get a warning message that only the first imputed data set was used. I assume this means that R2 was only calculated using the first imputed dataset. I asked about this on the [mc-stan discourse ](https://discourse.mc-stan.org/t/how-to-pool-bayesian-r-squared-values-in-multiple-imputation/17956) and it seems that R2 can be pooled from the combined posterior samples just like many other parameters. Would it be possible to add that functionality to the `bayes_R2()` method? I tried to look into doing it myself, but was struggling to understand the `fit$fit` S4 object.

```r
library(brms)
library(mice)
imp <- mice(nhanes2)
fit <- brm_multiple(bmi ~ age + hyp + chl, data = imp, chains = 1)
bayes_R2(fit_imp2)
#>    Estimate Est.Error      Q2.5    Q97.5
#> R2 0.4452458 0.1071932 0.2068827 0.623473
#> Warning message:
#> Using only the first imputed data set. Please interpret the results with caution until a more principled approach has been implemented. 
```

**Comment by jmgirard** [comment]
Hmm... it seems that `bayes_R2()` uses two primary variables: `ypred` and `y`.

`ypred` is created by `posterior_epred()` and `prepare_predictions()`. It looks like `prepare_predictions()` does include posterior `samples` from all imputed datasets for the model parameters but calculates the predictions using `x$data` which only includes the first imputed dataset. To pool this, I think we would need to align each sample with its original x data. Similarly, `y` is created by `get_y()` which only uses the first imputed dataset. 

A further complication is that it seems `brmsfit_multiple` objects only store the first imputed dataset in `$data`. Perhaps a workaround would be for the list of imputed dataframes (or mids object) to be given as an additional argument to a new `bayes_R2.brmsfit_multiple()` method.

**Comment by jmgirard** [comment]
Ok, I think the easiest approach is just to do this from the output of `brm_multiple()` when `combine = 0`.

```r
library(mice)
library(brms)

imp <- mice(nhanes2)

fit <- brm_multiple(
  formula = bmi ~ age + hyp + chl, 
  data = imp, 
  chains = 1, 
  combine = 0
)

pool_R2 <- function(object, probs = c(0.025, 0.975), ...) {
  r2_post <- vector(mode = "list", length = length(object))
  for (i in seq_along(object)) {
    r2_post[[i]] <- bayes_R2(object[[i]], summary = FALSE, ...)
  }
  r2_post_pooled <- do.call(rbind, r2_post)
  posterior_summary(r2_post_pooled, probs = probs)
}

pool_R2(fit)
#>    Estimate Est.Error     Q2.5     Q97.5
#> R2 0.456994 0.1516709 0.124549 0.6797989
```

Perhaps something like this function could be made into a method for a new class, for example: `bayes_R2.brm_multiple_uncombined()`.

**Comment by paul-buerkner** [comment]
I agree with the idea that we should add better post-processing support for `brm_multiple` models but this shouldn't just encompass `bayes_R2` but all post-processing methods making use of `.$data`. However, it is not yet clear to be what the best output should be once we start storing all data sets in the object. For example, should we just return a list of predictions and leave it to the user to merge them somehow or shall we do some automatic merging, and if yes, how shall that look like? 

I have changed the title of this issue to reflect the broader scope.

**Comment by jmgirard** [comment]
Awesome! I guess an important question would be this: Do you want to have a new class for brm_multiple output or do you want to somehow merge the models to fit the standard brmsfit class?

**Comment by al-obrien** [comment]
I like the brainstorming around how to best incorporate post-processing with `brm_multiple()`.  Not sure if this helps at all, but I thought I would share how I was dabbling with other model comparison metrics (waic, loo) and MICE: *would it also be possible to use the `combine = T` output from `brm_multiple()` and analyze the full imputed dataset as `newdata` in the various metrics?* 

Statistically speaking, I am *not* 100% confident if this is a correct approach but I thought I would share anyway. I tried using the complete set of posterior samples on a 'new' dataset containing all the imputations from MICE. My thinking was that this would ensure the log-likelihoods to calculate waic/loo were from both the complete set of posterior samples and imputed data they were generated from. If I am completely off-base or if this approach would be less than palatable, please let me know! 

## Setup MICE model and datasets
```r
library(mice)
library(brms)

# Create the impute object and the full set
imp <- mice(nhanes2)
imp_c <- mice::complete(imp, 'long')

mlist <- brm_multiple(
  formula = bmi ~ age + hyp + chl,
  warmup = 1000, 
  iter  = 2500,
  data = imp, 
  seed = 2020,
  chains = 2, 
  combine =T
)
```

## Model object post processing
```r
# With just the imputed dataset
brms::loo(mlist3)
```
```
Computed from 15000 by 25 log-likelihood matrix

         Estimate  SE
elpd_loo    -72.0 4.0
p_loo         9.6 2.0
looic       144.0 8.1
------
Monte Carlo SE of elpd_loo is NA.

Pareto k diagnostic values:
                         Count Pct.    Min. n_eff
(-Inf, 0.5]   (good)     17    68.0%   2193      
 (0.5, 0.7]   (ok)        6    24.0%   705       
   (0.7, 1]   (bad)       2     8.0%   421       
   (1, Inf)   (very bad)  0     0.0%   <NA>      
See help('pareto-k-diagnostic') for details.
Warning messages:
1: Using only the first imputed data set. Please interpret the results with caution until a more principled approach has been implemented. 
2: Found 2 observations with a pareto_k > 0.7 in model 'mlist3'. It is recommended to set 'moment_match = TRUE' in order to perform moment matching for problematic observations. 
```

## Model object post processing with complete dataset
```r
# With the complete imputed dataset by passing in as newdata
brms::loo(mlist3, newdata = imp_c)
```
```
Computed from 15000 by 125 log-likelihood matrix

         Estimate   SE
elpd_loo   -369.2 13.3
p_loo        57.3  9.2
looic       738.4 26.7
------
Monte Carlo SE of elpd_loo is NA.

Pareto k diagnostic values:
                         Count Pct.    Min. n_eff
(-Inf, 0.5]   (good)     89    71.2%   2193      
 (0.5, 0.7]   (ok)       30    24.0%   303       
   (0.7, 1]   (bad)       4     3.2%   111       
   (1, Inf)   (very bad)  2     1.6%   7         
See help('pareto-k-diagnostic') for details.
Warning message:
Found 6 observations with a pareto_k > 0.7 in model 'mlist3'. It is recommended to set 'moment_match = TRUE' in order to perform moment matching for problematic observations.  
```

**Comment by jmgirard** [comment]
@paul-buerkner If you can provide some guidance, I'd be happy to take a stab at this in a PR.

**Comment by paul-buerkner** [comment]
I agree with the options you layed out earlier. However, I didn't have enough time yet to have a good opinion on which approach is beneficial. Right now, I would prefer the structure being basically as is, that is, the object looking like a brmsfit object just with an additional child class `brmsfit_multiple` and, changing what is currently done right now, store all data sets in a list. The main decision problem I see is how to change the output of the methods. Most convenient would be just a list of output but there could be better options.

**Comment by hugoRW** [comment]
Hello,

On the same topic, when using the hdi function (from the bayestestR package), although there is no warning I suspect that the intervals displayed are based on the first imputed dataset as well - am I right? If so would there be an alternative reliable way to get the hdi (for a brm_mulitple model)?

Many thanks

H

**Comment by jmgirard** [comment]
> Hello,
> 
> On the same topic, when using the hdi function (from the bayestestR package), although there is no warning I suspect that the intervals displayed are based on the first imputed dataset as well - am I right? If so would there be an alternative reliable way to get the hdi (for a brm_mulitple model)?
> 
> Many thanks
> 
> H

When using the default `combine = TRUE`, the HDI function will summarize across the concatenated posteriors. So you should be fine.

**Comment by hugoRW** [comment]
> > Hello,
> > On the same topic, when using the hdi function (from the bayestestR package), although there is no warning I suspect that the intervals displayed are based on the first imputed dataset as well - am I right? If so would there be an alternative reliable way to get the hdi (for a brm_mulitple model)?
> > Many thanks
> > H
> 
> When using the default `combine = TRUE`, the HDI function will summarize across the concatenated posteriors. So you should be fine.


Many thanks for the ehlp. That was exactly what happened - I ran the model with default options ("combine=TRUE") and the hdi function wrked without any warning. Thanks a lot!
